File: go/tree/622_widthOfBinaryTree.go
package tree

import "math"

// widthOfBinaryTree二叉树最大宽度 BFS
// 假设当前节点的idx 是x 则，左子树是2x+1, 右子树是2x+2 (下标从0开始)
// 最大宽度就是每层的queue 中queue[last] - queue[0] + 1
func widthOfBinaryTree(root *TreeNode) int {
	if root == nil {
		return 0
	}

	width := math.MinInt32

	queue := make([]*Pair, 0)
	queue = append(queue, &Pair{0, root})

	for len(queue) > 0 {

		width = max(width, queue[len(queue)-1].idx-queue[0].idx+1)

		for _, v := range queue {
			// 此处循环的时候，有可能queue 里面通过下标queue[0], 或者queue[last] 获取的不仅仅是本层的元素
			// 虽然for range 遍历的queue 是固定的，但是通过下标获取的是最新的
			// width 的计算放到循环外面
			// width = max(width, queue[len(queue)-1].idx-queue[0].idx+1)
			queue = queue[1:]

			if v.node.Left != nil {
				// 此处的idx 应该为v的idx*2，而不是现有的queue 遍历时的i
				queue = append(queue, &Pair{2*v.idx + 1, v.node.Left})
			}

			if v.node.Right != nil {
				queue = append(queue, &Pair{2*v.idx + 2, v.node.Right})
			}

		}
	}

	return width
}

type Pair struct {
	idx  int
	node *TreeNode
}

File: rust/src/tree/622_width_of_binary_tree.rs

// width_of_binary_tree 最大深度，BFS
pub fn width_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {

    if root.is_none() {
        return 0
    }

    let mut width = 1;

    use std::collections::VecDeque;
    let mut queue = VecDeque::new();

    // queue 中存储的是元组，idx of node 和 node
    // 最大宽度则是计算每一层的queue[last].idx - queue[0].idx  + 1
    queue.push_back((0, root));

    while !queue.is_empty() {

        let first = queue.front().unwrap();
        let last = queue.back().unwrap();
        width = width.max(last.0 - first.0);

        for _ in 0..queue.len() {

            let head = queue.pop_front().unwrap();
            
            let idx = head.0;
            let node = head.1.unwrap();

            if node.borrow().left.is_some() {
                queue.push_back((idx * 2 + 1, node.borrow_mut().left.take()));
            }

            if node.borrow().right.is_some() {
                queue.push_back((idx * 2 + 2, node.borrow_mut().right.take()));
            }
        }

    }
    width
}